lib/exqlite/sqlite3.ex

```diff
@@ -110,25 +110,91 @@ defmodule Exqlite.Sqlite3 do
-      end
+  @doc """
+  Returns number of SQL parameters in a prepared statement.
+
+      iex> {:ok, conn} = Sqlite3.open(":memory:", [:readonly])
+      iex> {:ok, stmt} = Sqlite3.prepare(conn, "SELECT ?, ?")
+      iex> Sqlite3.bind_parameter_count(stmt)
+      2
+
+  """
+  @spec bind_parameter_count(statement) :: integer
+  def bind_parameter_count(stmt), do: Sqlite3NIF.bind_parameter_count(stmt)
+
+  @type bind_value ::
+          NaiveDateTime.t()
+          | DateTime.t()
+          | Date.t()
+          | Time.t()
+          | number
+          | iodata
+          | {:blob, iodata}
+          | atom
+
+  @deprecated "Use `bind/2` instead"
+  @spec bind(db, statement, [bind_value]) :: :ok
+  def bind(_conn, stmt, args), do: bind(stmt, args)
+
+  @doc """
+  Binds values to a prepared statement.
+
+      iex> {:ok, conn} = Sqlite3.open(":memory:", [:readonly])
+      iex> {:ok, stmt} = Sqlite3.prepare(conn, "SELECT ?, ?, ?, ?, ?")
+      iex> Sqlite3.bind(stmt, [42, 3.14, "Alice", {:blob, <<0, 0, 0>>}, nil])
+      iex> Sqlite3.step(conn, stmt)
+      {:row, [42, 3.14, "Alice", <<0, 0, 0>>, nil]}
+
+      iex> {:ok, conn} = Sqlite3.open(":memory:", [:readonly])
+      iex> {:ok, stmt} = Sqlite3.prepare(conn, "SELECT ?")
+      iex> Sqlite3.bind(stmt, [42, 3.14, "Alice"])
+      ** (ArgumentError) expected 1 arguments, got 3
+
+      iex> {:ok, conn} = Sqlite3.open(":memory:", [:readonly])
+      iex> {:ok, stmt} = Sqlite3.prepare(conn, "SELECT ?, ?")
+      iex> Sqlite3.bind(stmt, [42])
+      ** (ArgumentError) expected 2 arguments, got 1
+
+      iex> {:ok, conn} = Sqlite3.open(":memory:", [:readonly])
+      iex> {:ok, stmt} = Sqlite3.prepare(conn, "SELECT ?")
+      iex> Sqlite3.bind(stmt, [:erlang.list_to_pid(~c"<0.0.0>")])
+      ** (ArgumentError) unsupported type: #PID<0.0.0>
+
+  """
+  @spec bind(statement, [bind_value] | nil) :: :ok
+  def bind(stmt, nil), do: bind(stmt, [])
+
+  def bind(stmt, args) do
+    params_count = bind_parameter_count(stmt)
+    args_count = length(args)
+
+    if args_count == params_count do
+      bind_all(args, stmt, 1)
+    else
+      raise ArgumentError, "expected #{params_count} arguments, got #{args_count}"
+    end
   end
 
+  # credo:disable-for-next-line Credo.Check.Refactor.CyclomaticComplexity
+  defp bind_all([param | params], stmt, idx) do
+    case convert(param) do
+      i when is_integer(i) -> bind_integer(stmt, idx, i)
+      f when is_float(f) -> bind_float(stmt, idx, f)
+      b when is_binary(b) -> bind_text(stmt, idx, b)
+      b when is_list(b) -> bind_text(stmt, idx, IO.iodata_to_binary(b))
+      nil -> bind_null(stmt, idx)
+      :undefined -> bind_null(stmt, idx)
+      a when is_atom(a) -> bind_text(stmt, idx, Atom.to_string(a))
+      {:blob, b} when is_binary(b) -> bind_blob(stmt, idx, b)
+      {:blob, b} when is_list(b) -> bind_blob(stmt, idx, IO.iodata_to_binary(b))
+      _other -> raise ArgumentError, "unsupported type: #{inspect(param)}"
```

Comment:
I removed `BindError` to standardize on `ArgumentError`, since it was already used in `convert`.

Reply:
I added `BindError` because I could not figure out how to raise it directly from the NIF and I wanted to pass the term that had failed. I'm not attached to it.

> Maybe it could happen if blob or text are too large?

AFAIK this probably isn't possible under normal circumstances. You are allowed to redefine what the max limit of blobs can be when compiling, so you could in theory test this yourself locally by defining `-DSQLITE_MAX_LENGTH=4096` and doing a quick test to see the behavior.

https://www.sqlite.org/limits.html

Reply:
> raise it directly from the NIF 

AFAIK there is no way to do this right now, but there were some plans to add `error_info` to erl_nif APIs. `error_info` is already used in bifs and get's picked up by Elixir as well: https://github.com/elixir-lang/elixir/blob/be84111523cc18498b68949e4c2058eacb44fa00/lib/elixir/lib/exception.ex#L2483 I think that's how ets started returning not just "badarg" but

```elixir
iex> :ets.new "asdf", []
** (ArgumentError) errors were found at the given arguments:

  * 1st argument: not an atom

    (stdlib 6.0) :ets.new("asdf", [])
    iex:1: (file)
```